### PR TITLE
Enchance/refactor shourtcuts management

### PR DESCRIPTION
**Pull request by igoogolx**
Implement #447 .

### CODE REVIEW

File: app/renderer/services/shortcut-service.ts [hunk 1]
```diff
@@ -1,114 +1,69 @@
...
-          ) {
-            e.preventDefault();
-          }
-
-          this._registeredShortcutsInInputField[shortcut][handlerId]?.handler(
-            e
-          );
-
+      let shortcut = formatShortcut(e).join("+");
+      console.log(shortcut);
```
Comment (GeoffreyChen777) [inline app/renderer/services/shortcut-service.ts:35]: remove

**Comment by GeoffreyChen777** [comment]
We have a bug.

In the editing view's `<input>`s, `space` is not working.

I think it is due to:

https://github.com/Future-Scholars/paperlib/blob/618da2bf64a53a2cdd79b0b3feaec42dcefc488a/app/renderer/ui/main-view/main-view.vue#L44-L54

**Comment by igoogolx** [comment]
> We have a bug.
> 
> In all views' `<input>`s, `space` is not working.
> 
> I think it is due to:
> 
> https://github.com/Future-Scholars/paperlib/blob/618da2bf64a53a2cdd79b0b3feaec42dcefc488a/app/renderer/ui/main-view/main-view.vue#L44-L54

PreventDefault leads to this. I just fixed it by disabling preventDefault in input elements.

What do you think about [this](https://github.com/Future-Scholars/paperlib/pull/465/commits/85262c2202db0c486fddaf1a199dcebb9673bee1)?

**Comment by GeoffreyChen777** [comment]
One problem about `Meta`, `Control`, and `Command`.

Because in the new function `formatShortcut`, you convert them based on the OS.  However, in the previous versions, we set `CommandOrControl` as the default in the preference service. All shortcuts with `CommandOrControl` are not working now. Users need to reset all related shortcuts.

To solve it, I think we should increase `PREFERENCE_VERSION` in `preference-service.ts`, and in `_migrate`, convert those `CommandOrControl`.

**Comment by igoogolx** [comment]
> One problem about `Meta`, `Control`, and `Command`.
> 
> Because in the new function `formatShortcut`, you convert them based on the OS. However, in the previous versions, we set `CommandOrControl` as the default in the preference service. All shortcuts with `CommandOrControl` are not working now. Users need to reset all related shortcuts.
> 
> To solve it, I think we should increase `PREFERENCE_VERSION` in `preference-service.ts`, and in `_migrate`, convert those `CommandOrControl`.

`CommandOrControl ` should not be a problem because it is a default value but not **stored**  in preference. I have replaced all `CommandOrControl ` with `Command`  and `Control` in `preference-service.ts`.

**Comment by GeoffreyChen777** [comment]
image
If you print `_registeredShortcuts` you will see there are many `CommandOrControl`.

However, you convert shortcuts here:

```typescript
 window.addEventListener("keydown", (e) => {
      let shortcut = formatShortcut(e).join("+");           <-----
      console.log(shortcut);
      if (this._registeredShortcuts[shortcut]) {              <--- false
...
```

**Comment by igoogolx** [comment]
I see. We can do it in `formatShortcut`. BTW, I can't understand why `CommandOrControl `will be **stored** in preference.

**Comment by igoogolx** [comment]
> I see. We can do it in `formatShortcut`. BTW, I can't understand why `CommandOrControl `will be **stored** in preference.

I thought it is a default value. Only shortcuts inputted by users will be **stored**.

**Comment by GeoffreyChen777** [comment]
It's in the `_defaultPreferences` of all previous versions.

https://github.com/Future-Scholars/paperlib/blob/9262a3a5ea440d880f1e88ea4889a44235af63bd/app/common/services/preference-service.ts#L186-L194

**Comment by GeoffreyChen777** [comment]
> > I see. We can do it in `formatShortcut`. BTW, I can't understand why `CommandOrControl `will be **stored** in preference.
> 
> I thought it is a default value. Only shortcuts inputted by users will be **stored**.

Yes, it's default.

When you register shortcuts, you don't convert them. As a result, you register many `CommandOrControl` in `_registeredShortcuts `. But you convert shortcuts in `keydown event listener`.

**Comment by GeoffreyChen777** [comment]
BTW, if you just change the default value in `preference-service.ts`, it will not override users' config in the disk. You should do it explicitly in `_migrate`

https://github.com/Future-Scholars/paperlib/blob/9262a3a5ea440d880f1e88ea4889a44235af63bd/app/common/services/preference-service.ts#L217C10-L217C18

**Comment by igoogolx** [comment]
We can [convert ](https://github.com/Future-Scholars/paperlib/pull/465/commits/4327b170bc0839116ba0f4abb2f9c5f7ea59a570)`CommandOrControl` before registering shortcuts. So it is unnecessary to migrate.

**Comment by GeoffreyChen777** [comment]
I fixed some bugs. Now everything looks good.

Thanks for your contribution. Now we have a really neat shortcut management.

**Review by GeoffreyChen777 — APPROVED** [review]

**Pull request merged**